include/zephyr/kernel/obj_core.h

```diff
@@ -0,0 +1,353 @@
+#define __KERNEL_OBJ_CORE_H__
+
+#include <zephyr/sys/slist.h>
+
+/**
+ * @brief Macro to convert kernel object into pointer to object core
+ *
+ * This macro is used to convert a pointer to a kernel object (e.g. thread or
+ * memory block) into a pointer to its object core.
+ */
+#define K_OBJ_CORE(kobj)  (&((kobj)->obj_core))
+
+/* Known kernel object types */
+
+#define K_OBJ_TYPE_ID_GEN(s)     ((s[0] << 24) | (s[1] << 16) | (s[2] << 8) | (s[3]))
+
+#define K_OBJ_TYPE_CONDVAR_ID    K_OBJ_TYPE_ID_GEN("COND")
+#define K_OBJ_TYPE_CPU_ID        K_OBJ_TYPE_ID_GEN("CPU_")
+#define K_OBJ_TYPE_EVENT_ID      K_OBJ_TYPE_ID_GEN("EVNT")
+#define K_OBJ_TYPE_FIFO_ID       K_OBJ_TYPE_ID_GEN("FIFO")
+#define K_OBJ_TYPE_KERNEL_ID     K_OBJ_TYPE_ID_GEN("KRNL")
+#define K_OBJ_TYPE_LIFO_ID       K_OBJ_TYPE_ID_GEN("LIFO")
+#define K_OBJ_TYPE_MEM_BLOCK_ID  K_OBJ_TYPE_ID_GEN("MBLK")
+#define K_OBJ_TYPE_MBOX_ID       K_OBJ_TYPE_ID_GEN("MBOX")
+#define K_OBJ_TYPE_MEM_SLAB_ID   K_OBJ_TYPE_ID_GEN("SLAB")
+#define K_OBJ_TYPE_MSGQ_ID       K_OBJ_TYPE_ID_GEN("MSGQ")
+#define K_OBJ_TYPE_MUTEX_ID      K_OBJ_TYPE_ID_GEN("MUTX")
+#define K_OBJ_TYPE_PIPE_ID       K_OBJ_TYPE_ID_GEN("PIPE")
+#define K_OBJ_TYPE_SEM_ID        K_OBJ_TYPE_ID_GEN("SEM4")
+#define K_OBJ_TYPE_STACK_ID      K_OBJ_TYPE_ID_GEN("STCK")
+#define K_OBJ_TYPE_THREAD_ID     K_OBJ_TYPE_ID_GEN("THRD")
+#define K_OBJ_TYPE_TIMER_ID      K_OBJ_TYPE_ID_GEN("TIMR")
+
+struct k_obj_type;
+struct k_obj_core;
+
+#ifdef CONFIG_OBJ_CORE
+#define K_OBJ_CORE_INIT(_objp, _obj_type)   \
+	extern struct k_obj_type _obj_type; \
+	k_obj_core_init(_objp, &_obj_type)
+
+#define K_OBJ_CORE_LINK(objp) k_obj_core_link(objp)
+#else
+#define K_OBJ_CORE_INIT(objp, type)   do { } while (0)
+#define K_OBJ_CORE_LINK(objp)         do { } while (0)
+#endif
+/**
+ * Tools may use this list as an entry point to identify all registered
+ * object types and the object cores linked to them.
+ */
+extern sys_slist_t z_obj_type_list;
+
+struct k_obj_core_stats_desc {
+	size_t  raw_size;
+	size_t  query_size;
+
+	int (*raw)(struct k_obj_core *obj_core, void *stats);
+	int (*query)(struct k_obj_core *obj_core, void *stats);
+	int (*reset)(struct k_obj_core *obj_core);
+	int (*disable)(struct k_obj_core *obj_core);
+	int (*enable)(struct k_obj_core *obj_core);
+};
+
+struct k_obj_type {
+	sys_snode_t    node;   /* Node within list of object types */
+	sys_slist_t    list;   /* List of objects of this object type */
+	uint32_t       id;     /* Unique type ID */
+	size_t         obj_core_offset;  /* Offset to obj_core field */
+#ifdef CONFIG_OBJ_CORE_STATS
+	struct k_obj_core_stats_desc *stats_desc;
+#endif
+};
+
+struct k_obj_core {
+	sys_snode_t        node;   /* Object node within object type's list */
+	struct k_obj_type *type;   /* Object type to which object belongs */
+#ifdef CONFIG_OBJ_CORE_STATS
+	void  *stats;              /* Pointer to kernel object's stats */
+#endif
+};
+
+/**
+ * @defgroup obj_core_apis Object Core APIs
```

Comment:
Don't you want to move this further up? `k_obj_type`, `k_obj_core`, ... are API, I believe? Note how right now you don't have  them showing up in the index.rst page you created.

Also make sure to actually document everything that's API in this file, and hide what isn't behind `@cond INTERNAL_HIDDEN`.